4.5 - ORM\DeleteQuery shims

markstory opened the pull request:
Add a forwards compatibility implementation of the upcoming ORM\Query changes. My hope is that through deprecations and this new runtime interface enforcement solution will allow users to incrementally upgrade what I think could be an otherwise harsh break in compatibility.

This will create some merge conflicts with 5.x that I'm happy to take ownership of solving. If folks don't hate this approach I'll continue for all the other query types.

I've considered doing the same for the Database package to help our small but still important standalone package user base.

markstory committed:
First pass at DeleteQuery

This subclass of ORM\Query will enable userland code to adapt to the 5.x
API more easily. We can combine these new classes with deprecations to
encourage the use of new Table methods.

markstory committed:
Add Table::deleteQuery()

Soon, we'll be adding a deprecation to query() and update all of the
core usage of the ORM to use the new [verb]Query() methods.

This will result in all user-land code using query() directly to start
throwing errors.

markstory committed:
Add DeleteQuery

markstory committed:
Improve deprecation exception

Using a class in our package that has some additional formatting seems
wise.

othercorey commented on src/ORM/Exception/DeprecationException.php:
```diff
@@ -0,0 +1,14 @@
+    /**
+     * @inheritDoc
+     */
+    protected $_messageTemplate = 'Deprecated: %s';
```
The 5.x deprecation message includes the version number. Can we do it here too?

othercorey commented on src/ORM/Query/DeleteQuery.php:
```diff
@@ -0,0 +1,223 @@
+     */
+    public function delete(?string $table = null)
+    {
+        $this->_deprecatedException('delete()', 'Remove this method call.');
```
This isn't deprecated if we're throwing an exception. Somehow we need to communicate this call is not needed with the new DeleteQuery object.

markstory replied:
Should we be using regular deprecation warningsninstead of exceptions for along these methods then?

I was trying to make the required updates easy to find. However, using an exception will give a tougher upgrade than a warning would.

I will update these to use a warning instead.

markstory committed:
Make exceptions into deprecations

While using exceptions would give these changes high visibility, it also
is not backwards compatible and forces an early upgrade.

Using deprecationWarning() allows userland code to silence these errors
until they can be addressed.

Fix phpcs, psalm and tests.

markstory commented:
Last call on this. I'd like to keep making progress on the other query types.

markstory committed:
Add methods I missed earlier.

othercorey commented:
Looks ok

ADmad commented:
Looks good, just the phpcs/stan errors need to be addressed.

markstory committed:
Remove methods causing phpstan/psalm errors

othercorey commented:
patch has 7.1%, are you going to instantiate these shims somewhere?

markstory commented:
> are you going to instantiate these shims somewhere?

I've been trying to think of how to reasonably test the shims as well. Delete queries don't have all of these methods and wouldn't use the query parts if they were set. I could write tests that convert DeleteQuery into the other query types and ensure that DeleteQuery can still be coerced into an insert/update/select query.

markstory committed:
Add test case covering basic backwards compatibility interface

markstory commented:
@othercorey I added some rudimentary tests for commonly used methods and use cases. Are there others you'd like to see? I'm planning on having each deprecation emitting class go through the same tests via a provider.

othercorey commented:
> @othercorey I added some rudimentary tests for commonly used methods and use cases. Are there others you'd like to see? I'm planning on having each deprecation emitting class go through the same tests via a provider.

It's fine with me as long as people can instantiate the classes without fatal errors.
